### PR TITLE
Add Half support for reflection and replication padding on CPU

### LINKED ISSUE #135680: Meet an unimplemented crash for type 'Half' while the Data is not of Half type
### 🐛 Describe the bug

I meet a crash when using Keras with Pytorch backend and executing the following code.
```
import os
import re
import torch
import numpy as np
os.environ['KERAS_BACKEND']='torch'
import keras

layer = keras.layers.MaxPooling1D(
    pool_size=2,
    strides=1,
    padding="same",
    data_format="channels_first",
    trainable=True,
    dtype="mixed_float16",
    autocast=True,
)

result_dynamic = layer(
    inputs=np.random.rand(*[3, 5, 4]),
)
```
The error msg is:
```
...\torch\nn\functional.py:4552, in pad(input, pad, mode, value)
   4545         if mode == 'replicate':
   4546             # Use slow decomp whose backward will be in terms of index_put.
   4547             # importlib is required because the import cannot be top level
   4548             # (cycle) and cannot be nested (TS doesn't support)
   4549             return importlib.import_module('torch._decomp.decompositions')._replication_pad(
   4550                 input, pad
   4551             )
-> 4552 return torch._C._nn.pad(input, pad, mode, value)

RuntimeError: Exception encountered when calling MaxPooling1D.call().

"replication_pad1d" not implemented for 'Half'
```
The message is wired since the data is of dtype="mixed_float16" rather than 'Half'.

After reporting this [issue](https://github.com/keras-team/keras/issues/20236) to Keras team, the developers of Keras point out that it may be a problem of Pytorch backend, so I report this issue here.

### Versions

PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04 LTS (x86_64)
GCC version: (Ubuntu 13.2.0-23ubuntu4) 13.2.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:36:39) [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-5.15.153.1-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Quadro T1000
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      39 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             12
On-line CPU(s) list:                0-11
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
CPU family:                         6
Model:                              158
Thread(s) per core:                 2
Core(s) per socket:                 6
Socket(s):                          1
Stepping:                           10
BogoMIPS:                           5184.01
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid pni pclmulqdq ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt xsaveopt xsavec xgetbv1 xsaves md_clear flush_l1d arch_capabilities
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          192 KiB (6 instances)
L1i cache:                          192 KiB (6 instances)
L2 cache:                           1.5 MiB (6 instances)
L3 cache:                           12 MiB (1 instance)
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX unsupported
Vulnerability L1tf:                 Mitigation; PTE Inversion
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; IBRS, IBPB conditional, STIBP conditional, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Unknown: Dependent on hypervisor status
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] optree==0.11.0
[pip3] torch==2.4.0
[pip3] triton==3.0.0
[conda] numpy                     1.26.4                    <pip>
[conda] optree                    0.11.0                    <pip>
[conda] torch                     2.4.0                     <pip>
[conda] triton                    3.0.0                     <pip>

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### PR DESCRIPTION
**Pull request by CaoE**
Fixes #135680

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]
I was about to tackle these ops myself.

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -15214,8 +15214,7 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
            variant_test_name='reflect',
            supports_forward_ad=True,
            supports_fwgrad_bwgrad=True,
-           dtypes=all_types_and_complex_and(torch.bfloat16),
-           dtypesIfCUDA=all_types_and_complex_and(torch.half, torch.bfloat16),
+           dtypes=all_types_and_complex_and(torch.bfloat16, torch.half),
```
Comment (Skylion007) [inline torch/testing/_internal/common_methods_invocations.py:15217]: Need to add to `nn.functional.pad` as well.

Comment (CaoE) [reply torch/testing/_internal/common_methods_invocations.py:15217]: Thanks for the comments. Add it.

**Pull request closed without merging**